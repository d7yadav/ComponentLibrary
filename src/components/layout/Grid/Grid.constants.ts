import type { GridSize, GridDirection, GridWrap, GridJustify, GridAlign, GridSpacing } from './Grid.types';

export const GRID_SIZES: Record<string, GridSize> = {
  1: 1,
  2: 2,
  3: 3,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
  10: 10,
  11: 11,
  12: 12,
  auto: 'auto',
  true: true,
} as const;

export const GRID_DIRECTIONS: Record<GridDirection, GridDirection> = {
  row: 'row',
  'row-reverse': 'row-reverse',
  column: 'column',
  'column-reverse': 'column-reverse',
} as const;

export const GRID_WRAPS: Record<GridWrap, GridWrap> = {
  nowrap: 'nowrap',
  wrap: 'wrap',
  'wrap-reverse': 'wrap-reverse',
} as const;

export const GRID_JUSTIFY_CONTENT: Record<GridJustify, GridJustify> = {
  'flex-start': 'flex-start',
  center: 'center',
  'flex-end': 'flex-end',
  'space-between': 'space-between',
  'space-around': 'space-around',
  'space-evenly': 'space-evenly',
} as const;

export const GRID_ALIGN_ITEMS: Record<GridAlign, GridAlign> = {
  'flex-start': 'flex-start',
  center: 'center',
  'flex-end': 'flex-end',
  stretch: 'stretch',
  baseline: 'baseline',
} as const;

export const GRID_SPACING_VALUES: Record<GridSpacing, GridSpacing> = {
  0: 0,
  1: 1,
  2: 2,
  3: 3,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
  10: 10,
  11: 11,
  12: 12,
} as const;

export const GRID_BREAKPOINTS = {
  xs: 0,
  sm: 600,
  md: 900,
  lg: 1200,
  xl: 1536,
} as const;

export const GRID_COLUMN_COUNT = 12;

export const GRID_GUTTER_WIDTH = 24; // Default gutter width in pixels

export const AUTO_LAYOUT_CONFIGS = {
  cards: {
    minWidth: 280,
    maxWidth: 320,
    fit: true,
    fill: false,
  },
  tiles: {
    minWidth: 200,
    maxWidth: 250,
    fit: true,
    fill: false,
  },
  thumbnails: {
    minWidth: 120,
    maxWidth: 150,
    fit: true,
    fill: false,
  },
  responsive: {
    minWidth: 300,
    maxWidth: '1fr',
    fit: false,
    fill: true,
  },
} as const;

export const CSS_GRID_TEMPLATES = {
  // Common layout patterns
  twoColumn: 'repeat(2, 1fr)',
  threeColumn: 'repeat(3, 1fr)',
  fourColumn: 'repeat(4, 1fr)',
  autoFit: (minWidth: number | string) => `repeat(auto-fit, minmax(${minWidth}, 1fr))`,
  autoFill: (minWidth: number | string) => `repeat(auto-fill, minmax(${minWidth}, 1fr))`,
  
  // Responsive patterns
  responsive: {
    xs: '1fr',
    sm: 'repeat(2, 1fr)',
    md: 'repeat(3, 1fr)',
    lg: 'repeat(4, 1fr)',
    xl: 'repeat(6, 1fr)',
  },
  
  // Layout areas
  sidebar: `
    "header header"
    "sidebar main"
    "footer footer"
  `,
  hero: `
    "hero hero hero"
    "content content sidebar"
    "footer footer footer"
  `,
} as const;

export const ACCESSIBILITY_CONSTANTS = {
  containerRole: 'grid',
  itemRole: 'gridcell',
  listRole: 'list',
  listItemRole: 'listitem',
  regionRole: 'region',
  defaultLabel: 'Grid layout',
} as const;

export const RESPONSIVE_GRID_PATTERNS = {
  // Mobile-first responsive patterns
  mobileFirst: {
    xs: 12,
    sm: 6,
    md: 4,
    lg: 3,
    xl: 2,
  },
  
  // Desktop-first responsive patterns
  desktopFirst: {
    xs: 12,
    sm: 12,
    md: 6,
    lg: 4,
    xl: 3,
  },
  
  // Equal columns
  equal: {
    xs: 12,
    sm: 6,
    md: 6,
    lg: 6,
    xl: 6,
  },
  
  // Main content with sidebar
  mainSidebar: {
    main: {
      xs: 12,
      sm: 12,
      md: 8,
      lg: 9,
      xl: 9,
    },
    sidebar: {
      xs: 12,
      sm: 12,
      md: 4,
      lg: 3,
      xl: 3,
    },
  },
} as const;

export const DEFAULT_GRID_PROPS = {
  container: false,
  item: false,
  spacing: GRID_SPACING_VALUES[0],
  direction: GRID_DIRECTIONS.row,
  wrap: GRID_WRAPS.wrap,
  justifyContent: GRID_JUSTIFY_CONTENT['flex-start'],
  alignItems: GRID_ALIGN_ITEMS.stretch,
  autoFit: false,
  autoFill: false,
  useCssGrid: false,
} as const;