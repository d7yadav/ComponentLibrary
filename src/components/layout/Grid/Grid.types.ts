import type { GridProps as MuiGridProps } from '@mui/material/Grid';
import type { ReactNode } from 'react';

export type GridSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 'auto' | true;

export type GridDirection = 'row' | 'row-reverse' | 'column' | 'column-reverse';

export type GridWrap = 'nowrap' | 'wrap' | 'wrap-reverse';

export type GridJustify = 
  | 'flex-start' 
  | 'center' 
  | 'flex-end' 
  | 'space-between' 
  | 'space-around' 
  | 'space-evenly';

export type GridAlign = 
  | 'flex-start' 
  | 'center' 
  | 'flex-end' 
  | 'stretch' 
  | 'baseline';

export type GridSpacing = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export interface ResponsiveGridSize {
  xs?: GridSize;
  sm?: GridSize;
  md?: GridSize;
  lg?: GridSize;
  xl?: GridSize;
}

export interface GridProps extends Omit<MuiGridProps, 'container' | 'item' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'spacing' | 'direction' | 'wrap' | 'justifyContent' | 'alignItems'> {
  children?: ReactNode;
  
  container?: boolean;
  
  item?: boolean;
  
  xs?: GridSize;
  
  sm?: GridSize;
  
  md?: GridSize;
  
  lg?: GridSize;
  
  xl?: GridSize;
  
  spacing?: GridSpacing;
  
  rowSpacing?: GridSpacing;
  
  columnSpacing?: GridSpacing;
  
  direction?: GridDirection;
  
  wrap?: GridWrap;
  
  justifyContent?: GridJustify;
  
  alignItems?: GridAlign;
  
  alignContent?: GridAlign;
  
  minHeight?: number | string;
  
  bgcolor?: string;
  
  autoFit?: boolean;
  
  autoFill?: boolean;
  
  minColumnWidth?: number | string;
  
  maxColumnWidth?: number | string;
  
  columns?: number | string;
  
  rows?: number | string;
  
  gap?: number | string;
  
  rowGap?: number | string;
  
  columnGap?: number | string;
  
  templateAreas?: string;
  
  templateColumns?: string;
  
  templateRows?: string;
  
  useCssGrid?: boolean;
  
  className?: string;
  
  sx?: object;
  
  'aria-label'?: string;
  
  role?: string;
}

export interface GridStyleProps {
  container: boolean;
  item: boolean;
  xs?: GridSize;
  sm?: GridSize;
  md?: GridSize;
  lg?: GridSize;
  xl?: GridSize;
  spacing: GridSpacing;
  rowSpacing?: GridSpacing;
  columnSpacing?: GridSpacing;
  direction: GridDirection;
  wrap: GridWrap;
  justifyContent: GridJustify;
  alignItems: GridAlign;
  alignContent?: GridAlign;
  bgcolor?: string;
  minHeight?: number | string;
  autoFit: boolean;
  autoFill: boolean;
  minColumnWidth?: number | string;
  maxColumnWidth?: number | string;
  columns?: number | string;
  rows?: number | string;
  gap?: number | string;
  rowGap?: number | string;
  columnGap?: number | string;
  templateAreas?: string;
  templateColumns?: string;
  templateRows?: string;
  useCssGrid: boolean;
}

export interface GridBreakpoints {
  xs: number;
  sm: number;
  md: number;
  lg: number;
  xl: number;
}

export interface AutoLayoutConfig {
  minWidth: number | string;
  maxWidth: number | string;
  fit: boolean;
  fill: boolean;
}