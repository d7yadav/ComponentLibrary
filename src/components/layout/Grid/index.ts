/**
 * @fileoverview Grid component barrel export
 * @author Dilip Yadav <dilip.sm.yadav@gmail.com>
 */

export { Grid } from './Grid';
export type { 
  GridProps, 
  GridSize, 
  GridDirection,
  GridWrap,
  GridJustify,
  GridAlign,
  GridSpacing,
  ResponsiveGridSize,
  GridStyleProps,
  GridBreakpoints,
  AutoLayoutConfig
} from './Grid.types';
export {
  GRID_SIZES,
  GRID_DIRECTIONS,
  GRID_WRAPS,
  GRID_JUSTIFY_CONTENT,
  GRID_ALIGN_ITEMS,
  GRID_SPACING_VALUES,
  GRID_BREAKPOINTS,
  GRID_COLUMN_COUNT,
  GRID_GUTTER_WIDTH,
  AUTO_LAYOUT_CONFIGS,
  CSS_GRID_TEMPLATES,
  ACCESSIBILITY_CONSTANTS,
  RESPONSIVE_GRID_PATTERNS,
  DEFAULT_GRID_PROPS,
} from './Grid.constants';
export {
  CssGrid,
  AutoGrid,
  MasonryGrid,
  HolyGrailGrid,
  ResponsiveCardGrid,
} from './Grid.styles';