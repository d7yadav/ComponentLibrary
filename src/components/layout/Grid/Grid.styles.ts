import { Grid as MuiGrid } from '@mui/material';
import { styled } from '@mui/material/styles';
import React from 'react';

import { CSS_GRID_TEMPLATES } from './Grid.constants';
export const StyledGrid: React.ComponentType<any> = styled(MuiGrid, {
  shouldForwardProp: (prop) => !['customBgcolor', 'customMinHeight', 'customAutoFit', 'customAutoFill', 'customMinColumnWidth', 'customMaxColumnWidth', 'customColumns', 'customRows', 'customGap', 'customRowGap', 'customColumnGap', 'customTemplateAreas', 'customTemplateColumns', 'customTemplateRows', 'customUseCssGrid'].includes(prop as string),
})<any>((props: any) => {
  const {
    theme,
    customBgcolor,
    customMinHeight,
    customAutoFit,
    customAutoFill,
    customMinColumnWidth,
    customMaxColumnWidth,
    customColumns,
    customRows,
    customGap,
    customRowGap,
    customColumnGap,
    customTemplateAreas,
    customTemplateColumns,
    customTemplateRows,
    customUseCssGrid
  } = props;
  /**
   * Base styles for the grid container
   */
  const baseStyles = {
    // Background color
    ...(customBgcolor && {
      'backgroundColor': customBgcolor.includes('.')
        ? theme.palette[customBgcolor.split('.')[0] as keyof typeof theme.palette]?.main || customBgcolor
        : customBgcolor,
    }),
    // Minimum height
    ...(customMinHeight && {
      'minHeight': typeof customMinHeight === 'number' ? `${customMinHeight}px` : customMinHeight,
    }),
  };
  // CSS Grid specific styles
  if (customUseCssGrid) {
    return {
      ...baseStyles,
      'display': 'grid',
      // Grid template columns
      ...(customTemplateColumns && {
        'gridTemplateColumns': customTemplateColumns,
      }),
      // Auto-fit columns
      ...(customAutoFit && {
        'gridTemplateColumns': CSS_GRID_TEMPLATES.autoFit(customMinColumnWidth),
      }),
      // Auto-fill columns
      ...(customAutoFill && {
        'gridTemplateColumns': CSS_GRID_TEMPLATES.autoFill(customMinColumnWidth),
      }),
      // Simple column count
      ...(customColumns && {
        'gridTemplateColumns': typeof customColumns === 'number'
          ? `repeat(${customColumns}, 1fr)`
          : customColumns,
      }),
      // Grid template rows
      ...(customTemplateRows && {
        'gridTemplateRows': customTemplateRows,
      }),
      // Simple row count
      ...(customRows && {
        'gridTemplateRows': typeof customRows === 'number'
          ? `repeat(${customRows}, 1fr)`
          : customRows,
      }),
      // Grid template areas
      ...(customTemplateAreas && {
        'gridTemplateAreas': customTemplateAreas,
      }),
      // Gap properties
      ...(customGap && {
        gap: typeof customGap === 'number' ? theme.spacing(customGap) : customGap,
      }),
      ...(customRowGap && {
        rowGap: typeof customRowGap === 'number' ? theme.spacing(customRowGap) : customRowGap,
      }),
      ...(customColumnGap && {
        columnGap: typeof customColumnGap === 'number' ? theme.spacing(customColumnGap) : customColumnGap,
      }),
    };
  }
  // Default flexbox grid styles (MUI Grid behavior)
  return baseStyles;
});

// Specialized CSS Grid components - CssGrid component
export const CssGrid: React.ComponentType<any> = styled('div', {
  shouldForwardProp: (prop) => !['columns', 'rows', 'gap', 'minColumnWidth', 'autoFit', 'autoFill', 'templateAreas', 'responsive'].includes(prop as string),
})<{
  columns?: number | string,
  rows?: number | string,
  gap?: number | string,
  minColumnWidth?: number | string,
  autoFit?: boolean,
  autoFill?: boolean,
  templateAreas?: string,
  responsive?: boolean,
}>((props: any) => {
  const {
    theme,
    columns,
    rows,
    gap = 2,
    minColumnWidth = 250,
    autoFit = false,
    autoFill = false,
    templateAreas,
    responsive = false
  } = props;

  return {
    'display': 'grid',
    gap: typeof gap === 'number' ? theme.spacing(gap) : gap,
    // Auto layouts
    ...(autoFit && {
      'gridTemplateColumns': CSS_GRID_TEMPLATES.autoFit(
        typeof minColumnWidth === 'number' ? `${minColumnWidth}px` : minColumnWidth
      ),
    }),
    ...(autoFill && {
      'gridTemplateColumns': CSS_GRID_TEMPLATES.autoFill(
        typeof minColumnWidth === 'number' ? `${minColumnWidth}px` : minColumnWidth
      ),
    }),
    // Fixed columns
    ...(columns && {
      'gridTemplateColumns': typeof columns === 'number'
        ? `repeat(${columns}, 1fr)`
        : columns,
    }),
    // Fixed rows
    ...(rows && {
      'gridTemplateRows': typeof rows === 'number'
        ? `repeat(${rows}, 1fr)`
        : rows,
    }),
    // Template areas
    ...(templateAreas && {
      'gridTemplateAreas': templateAreas,
    }),
    // Responsive grid
    ...(responsive && {
      'gridTemplateColumns': CSS_GRID_TEMPLATES.responsive.xs,
      [theme.breakpoints.up('sm')]: {
        'gridTemplateColumns': CSS_GRID_TEMPLATES.responsive.sm,
      },
      [theme.breakpoints.up('md')]: {
        'gridTemplateColumns': CSS_GRID_TEMPLATES.responsive.md,
      },
      [theme.breakpoints.up('lg')]: {
        'gridTemplateColumns': CSS_GRID_TEMPLATES.responsive.lg,
      },
      [theme.breakpoints.up('xl')]: {
        'gridTemplateColumns': CSS_GRID_TEMPLATES.responsive.xl,
      },
    }),
  };
});

// Auto-layout grid for cards
export const AutoGrid: React.ComponentType<any> = styled(CssGrid)<{
  minWidth?: number | string,
  maxWidth?: number | string,
}>((props: any) => {
  const { minWidth = 280, maxWidth = '1fr' } = props;

  return {
    'gridTemplateColumns': `repeat(auto-fit, minmax(${typeof minWidth === 'number' ? `${minWidth}px` : minWidth}, ${typeof maxWidth === 'number' ? `${maxWidth}px` : maxWidth}))`,
  };
});

// Masonry-style grid
export const MasonryGrid: React.ComponentType<any> = styled('div', {
  shouldForwardProp: (prop) => !['columns', 'gap'].includes(prop as string),
})<{
  columns?: number,
  gap?: number,
}>((props: any) => {
  const { theme, columns = 3, gap = 2 } = props;

  return {
    columnCount: columns,
    columnGap: theme.spacing(gap),
    '& > *': {
      'breakInside': 'avoid',
      'marginBottom': theme.spacing(gap),
    },
    [theme.breakpoints.down('md')]: {
      columnCount: Math.max(1, columns - 1),
    },
    [theme.breakpoints.down('sm')]: {
      columnCount: Math.max(1, columns - 2),
    },
  };
});

// Holy Grail layout
export const HolyGrailGrid: React.ComponentType<any> = styled(CssGrid)((props: any) => {
  const { theme } = props;

  return {
    'gridTemplateAreas': CSS_GRID_TEMPLATES.hero,
    'gridTemplateColumns': '200px 1fr 200px',
    'gridTemplateRows': 'auto 1fr auto',
    'minHeight': '100vh',
    [theme.breakpoints.down('md')]: {
      'gridTemplateAreas': `
        "header"
        "main"
        "sidebar"
        "footer"
      `,
      'gridTemplateColumns': '1fr',
      'gridTemplateRows': 'auto 1fr auto auto',
    },
  };
});

// Responsive card grid
export const ResponsiveCardGrid: React.ComponentType<any> = styled(CssGrid)((props: any) => {
  const { theme } = props;

  return {
    'gridTemplateColumns': 'repeat(auto-fit, minmax(280px, 1fr))',
    gap: theme.spacing(3),
    [theme.breakpoints.down('sm')]: {
      'gridTemplateColumns': '1fr',
      gap: theme.spacing(2),
    },
    '&:hover': {
      'opacity': 0.8,
      'cursor': 'pointer',
    },
    '&:focus': {
      'outline': `2px solid ${theme.palette.primary.main}`,
      'outlineOffset': '2px',
    },
    '&:disabled': {
      'opacity': 0.5,
      'cursor': 'not-allowed',
      'pointerEvents': 'none',
    },
  };
});