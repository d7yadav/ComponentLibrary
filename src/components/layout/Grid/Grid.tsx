import React, { forwardRef, memo } from 'react';

import {
  DEFAULT_GRID_PROPS,
  ACCESSIBILITY_CONSTANTS,
} from './Grid.constants';
import { StyledGrid } from './Grid.styles';
import type { GridProps } from './Grid.types';

/**
 * Enhanced Grid component with advanced grid system and auto-layout support
 * 
 * Features:
 * - Flexbox grid system (MUI Grid) and CSS Grid support
 * - Responsive breakpoint support (xs, sm, md, lg, xl)
 * - Auto-fit and auto-fill column layouts
 * - CSS Grid template areas and custom layouts
 * - Flexible spacing options (gap, rowGap, columnGap)
 * - Direction and alignment controls
 * - Accessibility compliance (WCAG 2.1 AA)
 * - TypeScript support with strict typing
 * - Dark theme support via enhanced theme system
 */
export const Grid = forwardRef<HTMLDivElement, GridProps>((
  {
    children,
    container = DEFAULT_GRID_PROPS.container,
    item = DEFAULT_GRID_PROPS.item,
    xs,
    sm,
    md,
    lg,
    xl,
    spacing = DEFAULT_GRID_PROPS.spacing,
    rowSpacing,
    columnSpacing,
    direction = DEFAULT_GRID_PROPS.direction,
    wrap = DEFAULT_GRID_PROPS.wrap,
    justifyContent = DEFAULT_GRID_PROPS.justifyContent,
    alignItems = DEFAULT_GRID_PROPS.alignItems,
    alignContent,
    minHeight,
    bgcolor,
    autoFit = DEFAULT_GRID_PROPS.autoFit,
    autoFill = DEFAULT_GRID_PROPS.autoFill,
    minColumnWidth,
    maxColumnWidth,
    columns,
    rows,
    gap,
    rowGap,
    columnGap,
    templateAreas,
    templateColumns,
    templateRows,
    useCssGrid = DEFAULT_GRID_PROPS.useCssGrid,
    className,
    sx,
    'aria-label': ariaLabel,
    role,
    ...other
  },
  ref
) => {
  // Determine appropriate role based on usage
  const effectiveRole = role || (container ? ACCESSIBILITY_CONSTANTS.containerRole : ACCESSIBILITY_CONSTANTS.itemRole);
  
  // Accessibility props
  const accessibilityProps = {
    'aria-label': ariaLabel || (container ? ACCESSIBILITY_CONSTANTS.defaultLabel : undefined),
    role: effectiveRole,
  };

  return (
    <StyledGrid data-testid="grid"
      ref={ref}
      container={container}
      item={item}
      xs={xs}
      sm={sm}
      md={md}
      lg={lg}
      xl={xl}
      spacing={spacing}
      rowSpacing={rowSpacing}
      columnSpacing={columnSpacing}
      direction={direction}
      wrap={wrap}
      justifyContent={justifyContent}
      alignItems={alignItems}
      alignContent={alignContent}
      className={className}
      sx={sx}
      // Custom props for styled component
      customBgcolor={bgcolor}
      customMinHeight={minHeight}
      customAutoFit={autoFit}
      customAutoFill={autoFill}
      customMinColumnWidth={minColumnWidth}
      customMaxColumnWidth={maxColumnWidth}
      customColumns={columns}
      customRows={rows}
      customGap={gap}
      customRowGap={rowGap}
      customColumnGap={columnGap}
      customTemplateAreas={templateAreas}
      customTemplateColumns={templateColumns}
      customTemplateRows={templateRows}
      customUseCssGrid={useCssGrid}
      {...accessibilityProps}
      {...other}
    >
      {children}
    </StyledGrid>
  );
});

Grid.displayName = 'Grid';

export default memo(Grid);