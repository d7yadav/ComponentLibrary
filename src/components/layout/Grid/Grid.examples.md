# ðŸ“– Grid Examples

## Basic Usage

```typescript
import { Grid } from '@/components/layout/Grid';

// Basic usage
const Example = () => {
  return (
    <Grid>
      Basic Grid example
    </Grid>
  );
};
```

## Advanced Examples

```typescript
// Advanced configuration
const AdvancedExample = () => {
  const [value, setValue] = useState('');
  
  return (
    <Grid
      variant="primary"
      disabled={false}
      onChange={(e) => setValue(e.target.value)}
      sx={{
        borderRadius: 2,
        '&:hover': {
          transform: 'translateY(-2px)',
        },
      }}
    >
      Advanced Grid
    </Grid>
  );
};
```

## Integration Examples

```typescript
// Integration with form libraries
import { useController, Control } from 'react-hook-form';

interface FormIntegrationProps {
  name: string;
  control: Control;
}

const FormIntegration = ({ name, control }: FormIntegrationProps) => {
  const { field, fieldState } = useController({
    name,
    control,
    defaultValue: '',
  });

  return (
    <Grid
      {...field}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  );
};
```

## Edge Cases

```typescript
// Edge case: Empty states
const EmptyState = () => (
  <Grid disabled>
    No content available
  </Grid>
);

// Edge case: Error states
const ErrorState = () => (
  <Grid 
    error
    helperText="Something went wrong"
  />
);

// Edge case: Loading states
const LoadingState = () => (
  <Grid 
    loading
    disabled
  />
);
```

## Performance Optimization Examples

```typescript
// Memoized component for performance
const OptimizedGrid = React.memo((props: GridProps) => {
  const memoizedValue = useMemo(() => {
    // Expensive calculation
    return heavyComputation(props.value);
  }, [props.value]);

  return (
    <Grid 
      {...props}
      value={memoizedValue}
    />
  );
});

// Lazy loading for large lists
const LazyList = () => {
  return (
    <VirtualizedList
      itemCount={1000}
      renderItem={({ index }) => (
        <Grid key={index}>
          Item {index}
        </Grid>
      )}
    />
  );
};
```

## Accessibility Examples

```typescript
// Full accessibility implementation
const AccessibleExample = () => {
  return (
    <Grid
      aria-label="Descriptive label for screen readers"
      aria-describedby="helper-text"
      role="button"
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          handleAction();
        }
      }}
    >
      Accessible Grid
    </Grid>
  );
};

// High contrast mode support
const HighContrastExample = () => {
  return (
    <Grid
      sx={{
        '@media (prefers-contrast: high)': {
          borderWidth: 2,
          borderColor: 'text.primary',
        },
      }}
    >
      High contrast support
    </Grid>
  );
};
```

---

*These examples demonstrate real-world usage patterns and best practices for the Grid component.*
