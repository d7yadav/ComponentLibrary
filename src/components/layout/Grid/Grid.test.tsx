/**
 * @fileoverview Grid Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Grid } from './Grid';
import { GRID_CONSTANTS } from './Grid.constants';
import type { GridProps } from './Grid.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Grid Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Grid>Test content</Grid>);
      expect(screen.getByRole('grid')).toBeInTheDocument();
          });

    it('renders children correctly', () => {
      const testContent = 'Test Grid Content';
      renderWithTheme(<Grid>{testContent}</Grid>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-grid';
      renderWithTheme(
        <Grid data-testid={testId}>Test</Grid>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Grid className={customClass}>Test</Grid>
      );
      expect(screen.getByRole('grid')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: GridProps['variant'][] = [
        'container',
        'item',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Grid variant={variant} data-testid="variant-test">
              {variant} variant
            </Grid>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: GridProps['size'][] = [
        'auto',
        'true',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <Grid size={size} data-testid="size-test">
              {size} size
            </Grid>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });



    it('handles disabled state correctly', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Grid disabled onClick={handleClick} data-testid="disabled-test">
          Disabled Grid
        </Grid>
      );
      
      const element = screen.getByTestId('disabled-test');
      expect(element).toHaveAttribute('aria-disabled', 'true');
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles click events on grid container', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Grid onClick={handleClick} data-testid="clickable-grid">
          <Grid xs={12}>Grid Item</Grid>
        </Grid>
      );
      
      const element = screen.getByTestId('clickable-grid');
      fireEvent.click(element);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('maintains responsive behavior on interaction', () => {
      renderWithTheme(
        <Grid container spacing={2} data-testid="responsive-grid">
          <Grid xs={12} sm={6} md={4}>Item 1</Grid>
          <Grid xs={12} sm={6} md={4}>Item 2</Grid>
          <Grid xs={12} sm={6} md={4}>Item 3</Grid>
        </Grid>
      );
      
      const element = screen.getByTestId('responsive-grid');
      expect(element).toBeInTheDocument();
      expect(element).toHaveClass('MuiGrid-container');
    });
  });



  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Grid>Accessible Grid</Grid>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <Grid aria-label={ariaLabel}>
          Grid for screen readers
        </Grid>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<Grid>Keyboard Grid</Grid>);
      
      const element = screen.getByRole('grid');
      element.focus();
      expect(element).toHaveFocus();
          });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Grid 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA Grid
        </Grid>
      );
      
      const element = screen.getByRole('grid');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiGrid: {
            defaultProps: {
              variant: 'container',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Grid data-testid="themed-grid">
            Themed Grid
          </Grid>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-grid');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Grid>Dark Mode Grid</Grid>
        </ThemeProvider>
      );

      expect(screen.getByRole('grid')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <Grid key={i}>Grid {i}</Grid>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="grid"]')).toHaveLength(100);
    });

  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Grid />);
      expect(screen.getByRole('grid')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<Grid>{null}</Grid>);
      expect(screen.getByRole('grid')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Grid 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </Grid>
      );
      expect(screen.getByRole('grid')).toBeInTheDocument();
    });

    it('handles complex configurations', () => {
      renderWithTheme(
        <Grid
          variant="container"
          size="auto"
          disabled={false}
        >
          Complex Grid
        </Grid>
      );
      
      expect(screen.getByRole('grid')).toBeInTheDocument();
    });
  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Grid>Default Grid</Grid>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: GridProps['variant'][] = [
        'container',
        'item',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Grid variant={variant}>{variant} Grid</Grid>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
