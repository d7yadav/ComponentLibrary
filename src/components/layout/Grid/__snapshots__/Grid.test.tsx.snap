// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Grid Component > Snapshots > matches snapshot for all variants > variant-container 1`] = `
<div
  class="MuiGrid-root MuiGrid-direction-xs-row css-38ek44-MuiGrid-root"
  data-testid="grid"
  role="gridcell"
  variant="container"
>
  container
   Grid
</div>
`;

exports[`Grid Component > Snapshots > matches snapshot for all variants > variant-item 1`] = `
<div
  class="MuiGrid-root MuiGrid-direction-xs-row css-38ek44-MuiGrid-root"
  data-testid="grid"
  role="gridcell"
  variant="item"
>
  item
   Grid
</div>
`;

exports[`Grid Component > Snapshots > matches snapshot for default props 1`] = `
<div
  class="MuiGrid-root MuiGrid-direction-xs-row css-38ek44-MuiGrid-root"
  data-testid="grid"
  role="gridcell"
>
  Default Grid
</div>
`;
